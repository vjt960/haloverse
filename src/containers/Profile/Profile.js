import React from 'react';
import { connect } from 'react-redux';
import {
  CircularProgressbarWithChildren,
  buildStyles
} from 'react-circular-progressbar';
import 'react-circular-progressbar/dist/styles.css';
import './Profile.scss';

export class Profile extends React.Component {
  scrollToBottom = () => {
    this.loadPlayer.scrollIntoView({ behavior: 'smooth' });
  };

  componentDidMount() {
    this.scrollToBottom();
  }

  componentDidUpdate() {
    this.scrollToBottom();
  }

  getKDR = () => {
    const { stats } = this.props;
    const { kills, deaths } = stats;
    const total = kills + deaths;
    return (kills / total) * 100;
  };

  render() {
    const { user, spartan, emblem, stats } = this.props;
    const statusMeter = (
      <CircularProgressbarWithChildren
        value={this.getKDR()}
        strokeWidth={5}
        circleRatio={0.5}
        styles={buildStyles({
          strokeLinecap: 'butt',
          rotation: -0.25,
          trailColor: 'hsl(0, 0%, 93%)'
        })}
      >
        <img
          style={{ height: '100px', marginTop: '-5' }}
          src={emblem}
          alt={`${user}'s emblem`}
        />
        <div style={{ fontSize: '12', marginTop: '-5' }}>
          <p className="kdr-percentage">
            {(stats.kills / stats.deaths).toFixed(3)}
          </p>
          <p className="kdr-label">K/D</p>
        </div>
        <table className="statusMeter_table">
          <tbody>
            <tr>
              <td align="left">
                <span className="table-k">_</span>
                KILLS
              </td>
              <td align="right">{stats.kills}</td>
            </tr>
            <tr>
              <td align="left">
                <span className="table-d">_</span>
                DEATHS
              </td>
              <td align="right">{stats.deaths}</td>
            </tr>
          </tbody>
        </table>
        <p className="statusMeter_timer">{`${stats.time.hours}:${
          stats.time.minutes
        }:${Math.round(stats.time.seconds)}`}</p>
        <p className="statusMeter_playTime">PLAYTIME</p>
      </CircularProgressbarWithChildren>
    );
    return (
      <section className="profile_container">
        <h2 className="profile_gamertag">{user}</h2>
        <article className="profile_service-card">
          <table className="profile_stat-table" border="1">
            <tbody>
              <tr>
                <td>SPARTAN RANK</td>
                <td>{stats.rank}</td>
              </tr>
              <tr>
                <td>TOTAL XP</td>
                <td>{stats.xp}</td>
              </tr>
              <tr>
                <td>WINS</td>
                <td>{stats.wins}</td>
              </tr>
              <tr>
                <td>LOSSES</td>
                <td>{stats.losses}</td>
              </tr>
              <tr>
                <td>HEADSHOTS</td>
                <td>{stats.headshots}</td>
              </tr>
              <tr>
                <td>SHOTS FIRED</td>
                <td>{stats.shots}</td>
              </tr>
              <tr>
                <td>ACCURACY</td>
                <td>{`${((stats.hits / stats.shots) * 100).toFixed(3)}%`}</td>
              </tr>
              <tr>
                <td>DAMAGE DEALT</td>
                <td>
                  {stats.damage
                    .toFixed(2)
                    .toString()
                    .replace(/\B(?=(\d{3})+(?!\d))/g, ',')}
                </td>
              </tr>
              <tr>
                <td>MEDALS EARNED</td>
                <td>{stats.medals}</td>
              </tr>
            </tbody>
          </table>
          {statusMeter}
          <img
            className="profile_spartan-img"
            src={spartan}
            alt={`${user}'s Spartan`}
            ref={img => {
              this.loadPlayer = img;
            }}
          />
        </article>
      </section>
    );
  }
}

const mapStateToProps = state => ({
  user: state.user,
  spartan: state.spartan,
  emblem: state.emblem,
  stats: state.stats
});

export default connect(mapStateToProps)(Profile);
