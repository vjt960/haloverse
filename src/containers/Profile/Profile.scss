@import '../../sass/mixins.scss';
@import '../../sass/variables.scss';
@import url('https://fonts.googleapis.com/css?family=Orbitron&display=swap');

.profile_container {
  @include set-dimensions(auto, 100%);
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: center;
  background: hsl(0, 0%, 100%);
  background: linear-gradient(90deg, rgba(255,255,255,0) 0%, rgba(0,0,0,1) 80%, rgba(0,0,0,1) 100%);
  color: hsl(0, 0%, 100%);
}

.profile_gamertag {
  font-family: 'Orbitron', sans-serif;
  font-size: 2.25em;
  margin: 10px;
}

.profile_spartan-img {
  @include set-dimensions(100%, 100%);
  object-fit: contain;
}

.profile_service-card {
  @include set-dimensions(auto, 100%);
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-gap: 10px;
  padding: 10px;
}

.profile_emblem-img {
  @include set-dimensions(150px, 150px);
  object-fit: cover;
}

.statusMeter_table {
  width: 100%;
}

.profile_stat-table {
  @include set-dimensions(100%, 100%);
  color: black;
}

tr {
  text-align: center;
}

.table-k {
  background-color: #3E98C7;
  color: transparent;
  margin-right: 5px;
}

.table-d {
  background-color: hsl(0, 0%, 93%);
  color: transparent;
  margin-right: 5px;
}

.statusMeter_timer, 
.kdr-percentage {
  font-family: 'Orbitron', sans-serif;
  font-size: 2em;
  letter-spacing: 3px;
} 
.statusMeter_playTime, 
.kdr-label {
  font-family: 'Orbitron', sans-serif;
  font-size: 1.25em;
  text-align: center;
  letter-spacing: 3px;
}