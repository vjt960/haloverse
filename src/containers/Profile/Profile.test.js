import React from 'react';
import { shallow } from 'enzyme';
import { Profile, mapStateToProps } from './Profile';

describe('Profile', () => {
  describe('Profile Component', () => {
    let wrapper, instance;

    beforeEach(() => {
      const props = {
        user: 'exampleGamer',
        spartan: 'exampleURL',
        emblem: 'anotherURL',
        stats: {
          rank: 1,
          xp: 100,
          kills: 0,
          deaths: 0,
          wins: 0,
          losses: 0,
          headshots: 0,
          shots: 999,
          hits: 9,
          damage: 27,
          time: { hours: 5, minutes: 5, seconds: 5 },
          medals: 10
        }
      };
      wrapper = shallow(<Profile {...props} />, {
        disableLifecycleMethods: true
      });
      instance = wrapper.instance();
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should be able to get player Kill-Death ratio', () => {
      instance.getKDR = jest.fn();
      instance.getKDR();
      expect(instance.getKDR).toHaveBeenCalled();
    });
  });

  describe('mapStateToProps', () => {
    it('should return an object with user, spartan, emblem and stats from store', () => {
      const mockState = {
        user: 'exampleUser',
        spartan: 'exampleURL',
        emblem: 'anotherURL',
        stats: { stat: 'fact' },
        mockProp: 'fill',
        mockTwo: 'fill'
      };
      const expected = {
        user: 'exampleUser',
        spartan: 'exampleURL',
        emblem: 'anotherURL',
        stats: { stat: 'fact' }
      };

      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });
});
