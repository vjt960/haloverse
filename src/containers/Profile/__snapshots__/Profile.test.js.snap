// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Profile Component should match snapshot 1`] = `ShallowWrapper {}`;
